### PR TITLE
Support cuGraph-Service in cuGraph-PyG

### PR DESCRIPTION
**Pull request by alexbarghi-nv**
Resolves https://github.com/rapidsai/graph_dl/issues/33

This PR resolves the most critical issue for `cuGraph-PyG`, by adding support for `cuGraph-Service`.  `cuGraph-PyG` supports `cuGraph-Service` through the `RemoteGraph` API.  As a result of this API standardization, there are few major changes to `cugraph_store`, `dispatch`, and `cugraph_sampler`.  Instead, this PR primarily addresses some edge cases that only appear with remote graphs.

This PR also contains a bug fix for cuGraph-Service, where the `GraphVertexEdgeID` thrift union was not being created correctly, and an associated test update.  This bug was found during testing of cuGraph-Service+PyG.

This PR also contains a bug fix for cuGraph-Service, where cupy was mistakenly included as a required dependency.  It has now been made an optional (soft) dependency using `import_optional`.

This PR also adds a new notebook that demonstrates an end-to-end workflow using the PyG extensions and cuGraph-Service.

I also apologize for the large number of commits; this is due to the other PRs that this PR depends on not being merged.

### CODE REVIEW

**Review by eriknw — COMMENTED** [review]
I don't fully grok everything here, but the changes seem reasonable to me.

How much here should be or will be updated to use the new vector properties in PropertyGraph?

File: notebooks/gnn/cgs_creation_extensions/cgs_mag_extension.py [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+    last_offset = 0
+
+    for node_type, num_nodes in data[0]['num_nodes_dict'].items():
+        vertex_offsets[node_type] = last_offset
+        last_offset += num_nodes
+        
+        blank_df = cudf.DataFrame({'id':range(vertex_offsets[node_type], vertex_offsets[node_type] + num_nodes)})
+        blank_df.id = blank_df.id.astype('int64')
+        if isinstance(pG, MGPropertyGraph):
+            blank_df = dask_cudf.from_cudf(blank_df, npartitions=2)
```
[the remark this answers is not in the document]
Comment (alexbarghi-nv) [reply notebooks/gnn/cgs_creation_extensions/cgs_mag_extension.py:43]: the MAG dataset is different than the _other_ dataset we want to ingest with a similar name (I don't want to say the full name since this is a public repo)

**Comment by alexbarghi-nv** [comment]
Updating for vector properties is the eventual goal, but I don't want to do that during burndown.

File: notebooks/gnn/cgs_creation_extensions/cgs_mag_extension.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (alexbarghi-nv) [reply notebooks/gnn/cgs_creation_extensions/cgs_mag_extension.py:43]: 2 partitions is sufficient for MAG

File: notebooks/gnn/cgs_creation_extensions/cgs_mag_extension.py [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+        feature_df.columns = [str(c) for c in range(feature_df.shape[1])]
+        feature_df['id'] = range(vertex_offset, vertex_offset + node_features.shape[0])
+        feature_df.id = feature_df.id.astype('int64')
+        if isinstance(pG, MGPropertyGraph):
+            feature_df = dask_cudf.from_cudf(feature_df, npartitions=2)
+
+        pG.add_vertex_data(feature_df, vertex_col_name='id', type_name=node_type)
+
+    # Fill in an empty value for vertices without properties.
+    pG.fillna_vertices(0.0)
```
[the remark this answers is not in the document]
Comment (alexbarghi-nv) [reply notebooks/gnn/cgs_creation_extensions/cgs_mag_extension.py:60]: I would leave updating the fillna methods to after burndown.

[the remark this answers is not in the document]
Comment (alexbarghi-nv) [reply notebooks/gnn/cgs_creation_extensions/cgs_mag_extension.py:60]: Vector properties in cuGraph-PyG was always planned to be a 23.02 feature

File: python/cugraph-service/client/cugraph_service_client/remote_graph.py [hunk 3]
```diff
@@ -128,7 +128,9 @@ def _graph_id(self):
     def _client(self):
         return self.__client
 
-    def edges(self, backend=("cudf" if cudf is not None else "numpy")):
+    def edges(
+        self, backend=("cudf" if not isinstance(cudf, MissingModule) else "numpy")
```
[the remark this answers is not in the document]
Comment (alexbarghi-nv) [reply python/cugraph-service/client/cugraph_service_client/remote_graph.py:132]: clarified in docstring now

**Comment by alexbarghi-nv** [comment]
rerun tests

**Review by jnke2016 — APPROVED** [review]

**Comment by ChuckHastings** [comment]
rerun tests

**Comment by alexbarghi-nv** [comment]
rerun tests

File: notebooks/gnn/pyg_hetero_mag_cgs.ipynb [hunk 4]
```diff
@@ -0,0 +1,302 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": 1,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import rmm\n",
+    "\n",
+    "rmm.reinitialize(pool_allocator=True,initial_pool_size=5e+9, maximum_pool_size=20e+9)"
+   ]
```
Comment (VibhuJawa) [inline notebooks/gnn/pyg_hetero_mag_cgs.ipynb:30]: We should not create RMM Pool on a client on the service.  The only pool the client process should have is the Pytorch Caching Pool Allocator (enabled by default), having a pool here creates unnecessary memory  contention.

File: notebooks/gnn/pyg_hetero_mag_cgs.ipynb [hunk 5]
```diff
@@ -0,0 +1,302 @@
...
+    ")\n",
+    "\n",
+    "test_loader = NodeLoader(\n",
+    "    data=(feature_store, graph_store),\n",
+    "    shuffle=True,\n",
+    "    batch_size=50,\n",
+    "    node_sampler=sampler,\n",
+    "    input_nodes=('author', graph_store.get_vertex_index('author'))\n",
+    ")\n"
+   ]
```
Comment (VibhuJawa) [inline notebooks/gnn/pyg_hetero_mag_cgs.ipynb:140]: **Optional for now** 
Is the rest of code similar to `pyg_hetero_mag_cgs` ? If yes, i think it may make sense to put this in a model.py or something and just import for both notebooks ? 

I am not reviewing this assuming it  is the same code. Please let me know if it is not.  We can probably do this later on in a followup too as this PR seems to unblock the CI.

File: python/cugraph-service/scripts/default-config.sh [hunk 6]
```diff
@@ -20,19 +21,19 @@ THIS_DIR=$(cd $(dirname ${BASH_SOURCE[0]}) && pwd)
...
+export DASK_HOST_MEMORY_LIMIT=${DASK_HOST_MEMORY_LIMIT:-auto}
 
-BUILD_LOG_FILE=${BUILD_LOG_FILE:-${RESULTS_DIR}/build_log.txt}
-SCHEDULER_FILE=${SCHEDULER_FILE:-${WORKSPACE}/dask-scheduler.json}
-DATE=${DATE:-$(date --utc "+%Y-%m-%d_%H:%M:%S")_UTC}
-ENV_EXPORT_FILE=${ENV_EXPORT_FILE:-${WORKSPACE}/$(basename ${CONDA_ENV})-${DATE}.txt}
+export BUILD_LOG_FILE=${BUILD_LOG_FILE:-${RESULTS_DIR}/build_log.txt}
+export SCHEDULER_FILE=${SCHEDULER_FILE:-${WORKSPACE}/dask-scheduler.json}
+export DATE=${DATE:-$(date --utc "+%Y-%m-%d_%H:%M:%S")_UTC}
+export ENV_EXPORT_FILE=${ENV_EXPORT_FILE:-${WORKSPACE}/$(basename ${CONDA_ENV})-${DATE}.txt}
```
Comment (VibhuJawa) [inline python/cugraph-service/scripts/default-config.sh:39]: Why the exports ?

File: python/cugraph-service/client/cugraph_service_client/client.py [hunk 7]
```diff
@@ -1507,9 +1514,26 @@ def excepthook(exc):
...
+                id_or_ids = id_or_ids.get()
+            elif not isinstance(cudf, MissingModule) and isinstance(
+                id_or_ids, cudf.Series
+            ):
+                id_or_ids = id_or_ids.values_host
+            elif not isinstance(pandas, MissingModule) and isinstance(
+                id_or_ids, pandas.Series
+            ):
+                id_or_ids = id_or_ids.to_numpy()
+
```
Comment (rlratzel) [inline python/cugraph-service/client/cugraph_service_client/client.py:1529]: This can be a `FIXME` to get CI unblocked ASAP, but should there be a closing `else:` block to raise an informative exception for the unlikely event that none of the missing modules are present?

File: python/cugraph/cugraph/gnn/pyg_extensions/data/cugraph_store.py [hunk 8]
```diff
@@ -285,18 +288,26 @@ def is_multi_gpu(self):
         """
         return self.__graph.is_multi_gpu()
 
+    @cached_property
+    def is_remote(self):
+        return self.__graph.is_remote()
+
+    @cached_property
+    def _compute_required(self):
+        return self.is_multi_gpu and not self.is_remote
```
Comment (VibhuJawa) [inline python/cugraph/cugraph/gnn/pyg_extensions/data/cugraph_store.py:297]: **Nitpick:**
I really like the idea of having something like a `compute_required` but i am not sure about the name. Maybe something like a `_is_delayed_object` .  Feel free to ignore.

**Review by VibhuJawa — CHANGES_REQUESTED** [review]
Have requested some minor changes as well as minor changes.

File: python/cugraph-service/scripts/default-config.sh (hunk 6 above)
Comment (alexbarghi-nv) [reply python/cugraph-service/scripts/default-config.sh:39]: I didn't intend to add this to the PR, removing

File: python/cugraph-service/client/cugraph_service_client/remote_graph.py [hunk 9]
```diff
@@ -369,15 +423,40 @@ def get_edge_data(
         edge_ids=None,
         types=None,
         columns=None,
-        backend=("cudf" if cudf is not None else "numpy"),
+        backend=("cudf" if not isinstance(cudf, MissingModule) else "numpy"),
```
Comment (rlratzel) [inline python/cugraph-service/client/cugraph_service_client/remote_graph.py:426]: not critical at all, but would it be easier to define `cudf_installed = not(isinstance(cudf, MissingModule))` somewhere near the imports, then change code like this to:
```suggestion
        backend=("cudf" if cudf_installed else "numpy"),
```

File: python/cugraph/cugraph/gnn/pyg_extensions/data/cugraph_store.py (hunk 8 above)
Comment (alexbarghi-nv) [reply python/cugraph/cugraph/gnn/pyg_extensions/data/cugraph_store.py]: changed to `_is_delayed`

File: python/cugraph-service/client/cugraph_service_client/client.py [hunk 10]
```diff
@@ -1507,9 +1514,26 @@ def excepthook(exc):
...
+            if not isinstance(cp, MissingModule) and isinstance(id_or_ids, cp.ndarray):
+                id_or_ids = id_or_ids.get()
+            elif not isinstance(cudf, MissingModule) and isinstance(
+                id_or_ids, cudf.Series
+            ):
+                id_or_ids = id_or_ids.values_host
+            elif not isinstance(pandas, MissingModule) and isinstance(
+                id_or_ids, pandas.Series
+            ):
+                id_or_ids = id_or_ids.to_numpy()
```
[the remark this answers is not in the document]
Comment (alexbarghi-nv) [reply python/cugraph-service/client/cugraph_service_client/client.py:1528]: I don't think this is related to that bug.  I'm not indexing a dask dataframe here, just building a `GraphVertexEdgeID` object from a `np.ndarray` to send to the server

File: notebooks/gnn/pyg_hetero_mag_cgs.ipynb (hunk 4 above)
Comment (alexbarghi-nv) [reply notebooks/gnn/pyg_hetero_mag_cgs.ipynb]: resolved

File: python/cugraph-service/client/cugraph_service_client/client.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (VibhuJawa) [reply python/cugraph-service/client/cugraph_service_client/client.py:1528]: Ahh, okay.  Feel free to ignore it then.

File: notebooks/gnn/pyg_hetero_mag_cgs.ipynb (hunk 5 above)
Comment (alexbarghi-nv) [reply notebooks/gnn/pyg_hetero_mag_cgs.ipynb:111]: Code is similar but we may change the workflow a bit in the future.  Also because this is primarily for learning, the whole code should be in this file, I think

File: python/cugraph-service/scripts/default-config.sh [hunk 11]
```diff
@@ -12,6 +12,7 @@
 # limitations under the License.
 
 THIS_DIR=$(cd $(dirname ${BASH_SOURCE[0]}) && pwd)
+WORKSPACE=$(pwd)/..
```
Comment (rlratzel) [inline python/cugraph-service/scripts/default-config.sh]: It seems like this could break if run from the wrong directory. Can you make this relative to `THIS_DIR` if possible by default, and also make it override-able by the user if their environment has it defined? 
```suggestion
WORKSPACE=${WORKSPACE:${THIS_DIR}/..}
```

File: notebooks/gnn/pyg_hetero_mag_cgs.ipynb (hunk 5 above)
Comment (VibhuJawa) [reply notebooks/gnn/pyg_hetero_mag_cgs.ipynb:111]: Okay , then let it be as is.   Code duplication from a learning perspective makes sense.

File: python/cugraph-service/server/cugraph_service_server/cugraph_handler.py [hunk 12]
```diff
@@ -52,6 +58,8 @@
     GraphVertexEdgeIDWrapper,
 )
 
+ogb = import_optional("ogb")
```
Comment (rlratzel) [inline python/cugraph-service/server/cugraph_service_server/cugraph_handler.py:61]: I think you should be able to do this import in the extension instead.

**Review by rlratzel — APPROVED** [review]
Approving since my change requests are fairly minor and can probably be FIXMEs if necessary to get CI unblocked ASAP.

File: python/cugraph-service/client/cugraph_service_client/client.py (hunk 7 above)
Comment (alexbarghi-nv) [reply python/cugraph-service/client/cugraph_service_client/client.py:1533]: added one

File: python/cugraph-service/server/cugraph_service_server/cugraph_handler.py (hunk 12 above)
Comment (alexbarghi-nv) [reply python/cugraph-service/server/cugraph_service_server/cugraph_handler.py:61]: importing in the extension causes CGS to hang

File: python/cugraph-service/scripts/default-config.sh (hunk 11 above)
Comment (alexbarghi-nv) [reply python/cugraph-service/scripts/default-config.sh]: I didn't intend to add this change - this should be part of #2977

File: python/cugraph-service/client/cugraph_service_client/remote_graph.py (hunk 9 above)
Comment (alexbarghi-nv) [reply python/cugraph-service/client/cugraph_service_client/remote_graph.py]: done

File: python/cugraph-service/server/cugraph_service_server/cugraph_handler.py (hunk 12 above)
Comment (rlratzel) [reply python/cugraph-service/server/cugraph_service_server/cugraph_handler.py:61]: oh that's strange, we'll have to investigate that separately then.

**Comment by rlratzel** [comment]
@gpucibot merge

**Review by acostadon — COMMENTED** [review]
Learned a lot from reviewing,
I have nothing to add.

**Review by VibhuJawa — APPROVED** [review]
LGTM
